File: CPP/medium/combinationSum.cpp
#include <vector>
#include <algorithm>
#include <iostream>

void combineHelper(int target, int start, std::vector<int >candidates, std::vector<int> choice, std::vector<std::vector<int>>& result){
    if(target == 0){
        result.push_back(choice);
        return;
    }
    if(target < 0){
        return;
    }

    
    for(int i = start; i < candidates.size(); i++){
        choice.push_back(candidates[i]);
        combineHelper(target - candidates[i],i, candidates, choice, result);
        choice.pop_back();
        if(i != candidates.size() - 1 &&target < candidates[i + 1]){
            return;
        }
    }
}

std::vector<std::vector<int>> combinationSum(std::vector<int>& candidates, int target) {
    std::vector<int> temp = {};
    std::vector<std::vector<int>> result = {};
    std::sort(candidates.begin(), candidates.end());
    combineHelper(target,0, candidates, temp, result);
    return result;
}

// Test array   => {2,3,6,7}  Target = 7
// Combinations => {2,2,3}, {7}

// Test array   => {2,3,5}  Target = 8
// Combinations => {2,2,2,2}, {2,3,3}, {3,5}

// Test array   => {2}  Target = 1
// Combinations => {}

// Test array   => {1}  Target = 1
// Combinations => {1}

int main(){
    std::vector<std::pair<std::vector<int>, int>> testSet = {
        {{2,3,6,7},7},  {{2,3,5},8},
        {{2},1},{{1},1}
    };
    for(auto test : testSet){
        std::cout << "Test array   => {";
        for(int i = 0; i < test.first.size(); i++){
            std::cout << test.first[i] << (i == (test.first.size() - 1)? "}":",");
        }
        std::cout << "  Target = "<< test.second << std::endl;
        std::vector<std::vector<int>> answerSet = combinationSum(test.first, test.second);
        std::cout << "Combinations =>";
        if(answerSet.size() == 0) std::cout << " {}";
        for(int i = 0; i < answerSet.size(); i++){
            std::cout << " {";
            for(int j = 0; j < answerSet[i].size(); j++){
                std::cout << answerSet[i][j] << (j == (answerSet[i].size() - 1)? "}":",");
            }
            std::cout << (i == (answerSet.size() - 1)? "":",");
        }
        std::cout << std::endl <<std::endl; 
    }
    return 0;
}

File: Java/medium/combinationSum.java
import java.util.*;

class Solution {    
    public static void helper(int start, int[] candidates, int target, List<Integer> temp, List<List<Integer>> ans){
        if(target == 0){
            ans.add(new ArrayList<Integer>(temp));
        }
        if(target < 0){
            return;
        }

        for(int i = start; i < candidates.length; i++){
            temp.add(candidates[i]);
            helper(i, candidates, target - candidates[i], temp, ans);
            temp.remove(temp.size() - 1);
        }
    }

    public static List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        helper(0,candidates, target, temp, ans);

        return ans;
    }

    public static void printResult(List<List<Integer>> matrix){
        for(int i = 0; i < matrix.size(); i++){
            System.out.print(Arrays.toString( matrix.get(i).toArray()));
            if(i != matrix.size() - 1){
                System.out.print(",");
            } 
        }
    }

    // candidates = [2, 3, 6, 7], target = 7
    // Combination = [2, 2, 3],[7]

    // candidates = [2, 3, 5], target = 8
    // Combination = [2, 2, 2, 2],[2, 3, 3],[3, 5]

    // candidates = [2], target = 1
    // Combination =

    // candidates = [1], target = 1
    // Combination = [1]

    public static void main(String[] args){
        int[] targetSet = new int[]{7,8,1,1};
        int[][] testSet = new int[][]{{2,3,6,7}, {2,3,5}, {2}, {1}};
        for(int i = 0; i < testSet.length; i++){
            System.out.println("candidates = " + Arrays.toString(testSet[i]) + ", target = " + targetSet[i]);
            System.out.print("Combination = ");
            printResult(combinationSum(testSet[i], targetSet[i]));
            System.out.println('\n');
        }
    }
}